packages/next/src/server/app-render/dynamic-rendering.ts

```diff
@@ -0,0 +1,145 @@
+ * about the currently running code so that Next.js can make decisions on how to handle
+ * the current execution in different rendering modes such as pre-rendering, resuming, and SSR.
+ *
+ * Today Next.js treats all code as potentially static. Certain APIs may only make sense when dynamically rendering.
+ * Traditionally this meant deopting the entire render to dynamic however with PPR we can now deopt parts
+ * of a React tree as dynamic while still keeping other parts static. There are really two different kinds of
+ * Dynamic indications.
+ *
+ * The first is simply an intention to be dynamic. unstable_noStore is an example of this where
+ * the currently executing code simply declares that the current scope is dynamic but if you use it
+ * inside unstable_cache it can still be cached. This type of indication can be removed if we ever
+ * make the default dynamic to begin with because the only way you would ever be static is inside
+ * a cache scope which this indication does not affect.
+ *
+ * The second is an indication that a dynamic data source was read. This is a stronger form of dynamic
+ * because it means that it is inappropriate to cache this at all. using a dynamic data source inside
+ * unstable_cache should error. If you want to use some dynamic data inside unstable_cache you should
+ * read that data outside the cache and pass it in as an argument to the cached function.
+ */
+
+// Once postpone is in stable we should switch to importing the postpone export directly
+import React from 'react'
+
+import type { StaticGenerationStore } from '../../client/components/static-generation-async-storage.external'
+import { DynamicServerError } from '../../client/components/hooks-server-context'
+
+import { StaticGenBailoutError } from '../../client/components/static-generation-bailout'
+
+const hasPostpone = typeof React.unstable_postpone === 'function'
+
+/**
+ * This function communicates that the current scope should be treated as dynamic.
+ *
+ * In most cases this function is a no-op but if called during
+ * a PPR prerender it will postpone the current sub-tree.
+ */
+export function markCurrentScopeAsDynamic(
+  store: StaticGenerationStore,
+  expression: string
+): void {
+  if (store.isUnstableCacheCallback) {
+    // inside cache scopes marking a scope as dynamic has no effect because the outer cache scope
+    // creates a cache boundary. This is subtly different from reading a dynamic data source which is
+    // forbidden inside a cache scope.
+    return
+  } else if (store.dynamicShouldError) {
+    throw new StaticGenBailoutError(
+      `Page with \`dynamic = "error"\` couldn't be rendered statically because it used \`${expression}\`. See more info here: https://nextjs.org/docs/app/building-your-application/rendering/static-and-dynamic#dynamic-rendering`
+    )
+  } else if (
+    // We are in a prerender (PPR enabled, during build)
+    store.prerenderState &&
+    hasPostpone
```

Comment:
That's a good point. It's a shame it needs to be a runtime check but until postpone is part of canary I think we're stuck with this. It is possible that it would be undefined if we messed up the React version loading so I've put in a new assertion that actually tests this. I put it in the conditional b/c I was thinking it was effectively required to be true and that would allow TS to refine the type in that scope however I was wrong about it being an impossible state. Nice find, this could have been an enormously frustrating bug to track down in the future... :)